File: 628.maximum-product-of-three-numbers.cpp
/*
 * @lc app=leetcode id=628 lang=cpp
 *
 * [628] Maximum Product of Three Numbers
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int maximumProduct(vector<int>& nums) {
    int max1 = INT_MIN;
    int max2 = INT_MIN;
    int max3 = INT_MIN;
    int min1 = INT_MAX;
    int min2 = INT_MAX;
    for (const auto num : nums) {
        if (num > max1) {
            max3 = max2;
            max2 = max1;
            max1 = num;
        } else if (num > max2) {
            max3 = max2;
            max2 = num;
        } else if (num > max3) {
            max3 = num;
        }
        if (num < min1) {
            min2 = min1;
            min1 = num;
        } else if (num < min2) {
            min2 = num;
        }
    }
    return std::max(max1 * max2 * max3, max1 * min1 * min2);
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  nums = {1, 2, 3};
  ASSERT_EQ(s.maximumProduct(nums), 6);
  nums = {1, 2, 3, 4};
  ASSERT_EQ(s.maximumProduct(nums), 24);
  nums = {-1, -2, -3};
  ASSERT_EQ(s.maximumProduct(nums), -6);

  return 0;
}

File: 628.maximum-product-of-three-numbers.rs
/*
 * @lc app=leetcode id=628 lang=rust
 *
 * [628] Maximum Product of Three Numbers
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn maximum_product(mut nums: Vec<i32>) -> i32 {
        nums.sort_unstable();
        let len = nums.len();
        std::cmp::max(
            nums[len - 1] * nums[len - 2] * nums[len - 3],
            nums[len - 1] * nums[0] * nums[1],
        )
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::maximum_product(vec![1, 2, 3]), 6);
}

#[test]
fn test2() {
    assert_eq!(Solution::maximum_product(vec![1, 2, 3, 4]), 24);
}

#[test]
fn test3() {
    assert_eq!(Solution::maximum_product(vec![-1, -2, -3]), -6);
}
